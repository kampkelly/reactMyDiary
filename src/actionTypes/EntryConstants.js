export const NEW_ENTRY = 'NEW_ENTRY';
