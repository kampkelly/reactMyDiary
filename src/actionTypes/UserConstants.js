export const SIGNIN = 'SIGNIN';
export const SIGNUP = 'SIGNUP';
export const SIGNOUT = 'SIGNOUT';
export const PROFILE = 'PROFILE';
export const NOTIFICATION = 'NOTIFICATION';