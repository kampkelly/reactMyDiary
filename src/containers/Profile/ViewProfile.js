import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';

import './ViewProfile.scss';
import '../../styles/sidebar.scss';
import { SaveNotification, ShowProfile } from '../../requests/UserRequests';
import { AllEntries } from '../../requests/EntryRequests';
import { validateForm } from '../../helpers/validateForm';
import icon from '../../assets/Rolling.svg';

/**
 * @class ViewEntry
 * @extends {Component}
 */
class ViewProfile extends Component {
  /**
   *Creates an instance of ViewProfile.
   * @memberof ViewProfile
   */
  constructor() {
    super();
    this.state = {
      diaryEntries: 0,
      entry: {},
      mmessage: '',
      reminderTime: '__:__',
      success: false,
      user: {}
    };
    this.change = this.change.bind(this);
    this.changeSettings = this.changeSettings.bind(this);
    this.saveSettings = this.saveSettings.bind(this);
  }

  change(e) {
    this.setState({ [e.target.name]: e.target.value });
  }

  checkNotice() {
    const pageUrl = window.location.href;
    const url = new URL(pageUrl);
    const notice = url.searchParams.get('notice');
    if (notice) {
      if (notice.length >= 1) {
        document.querySelector('#flash-message').style.display = 'block';
        document.querySelector('#flash-message p').textContent = notice;
        if (url.searchParams.get('warning')) {
          document.querySelector('#flash-message').style.backgroundColor = '#e00a1e';
        }
      }
    }
  }

  componentDidMount() {
    this.props.ShowProfile();
    this.props.AllEntries();
    this.checkNotice();
    document.querySelector('body').insertAdjacentHTML('afterbegin', `<img src=${icon} id="loading"></img>`);
  }

  /**
   *
   *
   * @param {*} e
   * @memberof ViewProfile
   */
  changeSettings(e) {
    e.preventDefault();
    document.querySelectorAll('#profile #settings form input[type="time"]')[0].removeAttribute('disabled');
    document.querySelectorAll('#profile #settings form button')[0].removeAttribute('disabled');
    document.querySelectorAll('#profile #settings form input[type="time"]')[0].style.backgroundColor = 'white';
    document.querySelectorAll('#profile #settings form button')[0].style.backgroundColor = '#052F60';
    document.querySelectorAll('#profile #settings form button')[0].style.color = 'white';
    document.querySelectorAll('#profile #settings form button')[0].innerHTML = 'Save';
    // document.querySelectorAll('#profile #settings form button')[0].addEventListener('mouseover', function () {
    //   document.querySelectorAll('#profile #settings form button')[0].style.backgroundColor = '#8c8c8c';
    // });
    // document.querySelectorAll('#profile #settings form button')[0].addEventListener('mouseleave', function () {
    //   document.querySelectorAll('#profile #settings form button')[0].style.backgroundColor = '#052F60';
    // });
  }

  /**
   * @description - This method runs whenever redux state changes
   * @returns {object} state
   * @param {object} props
   * @param {object} state
   * @memberof ViewProfile
   */
  static getDerivedStateFromProps(props, state) {
    return {
      diaryEntries: props.entries.length,
      user: props.user,
      reminderTime: props.user.remindertime,
      success: props.success
    };
  }


  saveSettings(e) {
    e.preventDefault();
    const errorMessage = 'Please Enter a Valid Date!';
    const values = { reminderTime: this.state.reminderTime };
    const noneEmpty = validateForm(errorMessage, values, e);
    if (noneEmpty === true) {
      document.querySelector('body').insertAdjacentHTML('afterbegin', `<img src=${icon} id="loading"></img>`);
      this.props.SaveNotification(this.state.reminderTime);
    }
  }

  /**
   *
   * @returns {jsx} - jsx
   * @memberof ViewProfile
   */
  showProfile() {
    let dateofbirth = '';
    if (this.state.user.dateofbirth && this.state.user.dateofbirth !== ' ') {
			dateofbirth = this.state.user.dateofbirth.split('T')[0];
		}
    if (this.state.user.email) {
      document.getElementById('loading').style.display = 'none';
      const createdat = this.state.user.createdat.split('T')[0];
      return (<div><ul className="no-styling"><li ><strong>Email:</strong>{this.state.user.email}</li><li ><strong>Name:</strong>{this.state.user.fullname}</li><li ><strong>Date of birth:</strong>{dateofbirth}</li><li ><strong>Date Joined:</strong>{createdat}</li></ul>
      <div className=""><Link to="profile/edit" className="underline"><small>Edit Profile</small></Link><p className="light-white-text"><small>Total number of entries in diary: {this.state.diaryEntries}</small></p></div></div>);
    } else {
      return (<p>user is not loaded</p>);
    }
  }

  /**
   * @description - This method renders the jsx for this component
   * @returns {jsx} - jsx
   * @memberof ViewProfile
   */
  render() {
    return (
      <div id="profile">
        <h2 className="underline title text-center">My Profile</h2>
	        <div className="grid-container">
							<div className="col-3-6 col-6-6-xs col-6-6-md">
								<section id="profile_details">
		              <h4 className="underline title text-center">Details</h4>
                  {this.showProfile()}
									<div id="insert"></div>
									<p className="danger-text notification_text"><small>Reminder has been set to: <span id="notification_time">{this.state.reminderTime}__:__</span> every day.</small></p>
		            </section>
							</div>
							<div className="col-3-6 col-6-6-xs col-6-6-md mt-4-md mt-4-xs">
								<section id="settings">
									<div className="text-center">
										<h4 className="underline title text-center">Notification Setting</h4>
										<a href="#" id="change_settings" className="underline" onClick={this.changeSettings}><small>Click here to change notification setting</small></a>
									</div>
                    <div className="grid-container mt-2">
                        <div className="col-6-6 col-6-6-md col-6-6-xs">
                          <form className="signin_form blue-forms form-cover" id="signin-form">
                            <p className="danger-text form_error_text"><small></small></p>
                            <div className="field grid-container">
                              <div className="col-6-6 col-6-6-md col-6-6-xs">
                                  <label className="white-text">Date:</label>
                              </div>
                              <div className="col-6-6 col-6-6-md col-6-6-xs">
                                  <input type="time" required="true" className="control-form" name="reminderTime" id="settngs" placeholder="Enter Date" autoComplete="settings" disabled="true" onChange={this.change} defaultValue={this.state.reminderTime} />
                              </div>
                            </div>
                            <div className="field grid-container">
                              <div className="col-2-6 col-6-6-md col-6-6-xs"></div>
                              <div className="col-6-6 col-6-6-md col-6-6-xs">
                                  <button type="submit" className="submit-button button-white" onClick={this.saveSettings} >Edit</button>
                              </div>
                            </div>
                          </form>
                        </div>
                    </div>
                </section>
						</div>
					</div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  entries: state.entry.entries,
  user: state.user.user,
  success: state.user.success
});

ViewProfile.propTypes = {
  AllEntries: PropTypes.func,
  SaveNotification: PropTypes.func,
  ShowProfile: PropTypes.func,
  success: PropTypes.bool
};

export default connect(mapStateToProps, {
  AllEntries, SaveNotification, ShowProfile
})(ViewProfile);

