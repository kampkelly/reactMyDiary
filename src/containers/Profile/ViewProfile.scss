$primary-color: #052F60;
$secondary-color: #DFAC2C;
$light-white-color: #8c8c8c;
$white-color: #fafafa;
$smartphone: "(min-width: 0px) and (max-width: 599px)";
$tablet: "(min-width: 600px) and (max-width: 900px)";

main {
	// background-image: url("https://i.imgur.com/n4ttyU5.jpg");
  background-size: cover;
	background-position: center;
	background-repeat-x: no-repeat;
  background-repeat-y: no-repeat;
}
#profile {
	padding-top: 30px;
	width: 90%;
	margin: auto;
    ul {
        li {
            margin-bottom: 10px;
        }
    }
	.form-cover {
        background: rgba(107, 101, 101, 0.3);
        margin: auto;
    }
}
#profile_details {
	background: #e1e1e1;
	padding: 30px;
	min-height: 350px;
	li, p {
		color: #DEAB3B;
	}
	p.notification_text {
		color: red;
	}
}
#settings {
	background: #c8c8c8;
	padding: 30px;
	min-height: 350px;
    form {
        color: gray;
        padding: 10px;
		width: 80%;
		margin: auto;
        .field {
            background: none;
            .control-form {
                width: 100%;
                background: gray;
            }
            button {
                width: 100%;
                background: $light-white-color;
            }
        }
    }
}
@media #{$smartphone} {
    #profile_details, #settings {
        // padding-left: 20px;
    }
	form {
		width: 90%;
	}
}
@media #{$tablet} {
    #settings form {
        margin: auto;
        width: 70%;
    }
	[class*='col-'] {
	  padding-right: 0px;
	}
}
