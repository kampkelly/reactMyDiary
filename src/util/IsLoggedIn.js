export const isLoggedIn = () => {
  if (localStorage.getItem('diary_token') !== null) {
    return true;
  }
  return false;
};
