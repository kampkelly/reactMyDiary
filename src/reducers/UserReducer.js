import { SIGNIN, SIGNUP, SIGNOUT, PROFILE, NOTIFICATION, UPDATE_PROFILE } from '../actionTypes/UserConstants';
import { asyncActionName } from '../util/AsyncUtil';


const initialState = {
  isAuth: false,
  loading: false,
  message: '',
  success: false,
  failure: false,
  user: {}
};

const UserReducer = (state = initialState, action = {}) => {
  switch (action.type) {
    case asyncActionName(SIGNIN).loading:
      return {
        ...state,
        loading: action.payload,
      };
    case asyncActionName(SIGNIN).success:
      return {
        ...state,
        success: true, ...action.payload.user, isAuth: true
      };
    case asyncActionName(SIGNIN).failure:
      document.querySelector('.form_error_text').style.display = 'block';
      document.querySelector('.form_error_text small').textContent = action.payload.error;
      return {
        ...state,
      };
    case asyncActionName(SIGNUP).loading:
      return {
        ...state,
        loading: action.payload,
      };
    case asyncActionName(SIGNUP).success:
      return {
        ...state,
        success: action.payload,
      };
    case asyncActionName(SIGNUP).failure:
      document.querySelector('.form_error_text').style.display = 'block';
      document.querySelector('.form_error_text small').textContent = action.payload.error;
      return {
        ...state,
      };
    case asyncActionName(PROFILE).loading:
      return {
        ...state,
        loading: action.payload,
      };
    case asyncActionName(PROFILE).success:
      return {
        ...state,
        success: true, user: action.payload
      };
    case asyncActionName(PROFILE).failure:
      return {
        ...state,
      };
    case asyncActionName(UPDATE_PROFILE).loading:
      return {
        ...state,
        loading: action.payload,
      };
    case asyncActionName(UPDATE_PROFILE).success:
      return {
        ...state,
        message: action.payload.message, success: true, user: action.payload.user
      };
    case asyncActionName(UPDATE_PROFILE).failure:
      document.querySelector('.form_error_text').style.display = 'block';
			document.querySelector('.form_error_text small').textContent = data.message;
      return {
        ...state,
      };
    case asyncActionName(NOTIFICATION).loading:
      return {
        ...state,
        loading: action.payload,
      };
    case asyncActionName(NOTIFICATION).success:
      return {
        ...state,
        success: true, message: action.payload.message, user: action.payload.user
      };
    case asyncActionName(NOTIFICATION).failure:
      return {
        ...state,
      };
    case asyncActionName(SIGNOUT).success:
      return initialState;
    default:
      return state;
  }
};

export default userReducer;
