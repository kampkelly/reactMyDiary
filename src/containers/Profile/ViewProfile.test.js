import React from 'react';
import { shallow } from 'enzyme';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import ViewProfile from './ViewProfile';
import UserReducer from '../../reducers/UserReducer';
import { asyncActions } from '../../util/AsyncUtil';
import { PROFILE } from '../../actionTypes/UserConstants';

const mockStore = configureMockStore([thunk]);
const store = mockStore({
  user: {
    user: {},
    message: '',
    success: false
  },
  entry: {
    entries: []
  }
});

const props = {
  history: {
    push: jest.fn()
  },
  checkNotice: jest.fn(),
  changeSettings: jest.fn(),
  saveSettings: jest.fn(),
  showProfile: jest.fn()
};

let component;
let myComponent;

describe('<ViewProfile/>', () => {
  beforeEach(() => {
    component = shallow(
      <Provider store={store}>
        <ViewProfile {...props} />
      </Provider>
    );
    myComponent = component.dive({ context: { store } }).dive();
  });
  it('should render without throwing an error', () => {
    expect(component).toMatchSnapshot();
  });
  it('should show profile', () => {
    const payload = {};
    const expectedAction = {
      type: 'PROFILE_SUCCESS',
      payload
    };
    expect(asyncActions(PROFILE).success(payload)).toEqual(expectedAction);
  });
  it('should dispatch an action when showing profile is loading', () => {
    const payload = false;
    const action = {
      type: 'PROFILE_LOADING',
      payload
    };
    const newState = UserReducer({}, action);
    expect(newState).toEqual({
      loading: payload
    });
  });
  it('should dispatch an action to show a profile', () => {
    const payload = {};
    const action = {
      type: 'PROFILE_SUCCESS',
      payload
    };
    const newState = UserReducer({}, action);
    expect(newState).toEqual({
      user: action.payload,
      success: true,
    });
  });
  it('should dispatch an action when updating an entry fails', () => {
    const payload = {};
    const action = {
      type: 'PROFILE_FAILING',
      payload
    };
    const newState = UserReducer({}, action);
    expect(newState).toEqual({});
  });
  it('should dispatch an action when notification is loading', () => {
    const payload = false;
    const action = {
      type: 'NOTIFICATION_LOADING',
      payload
    };
    const newState = UserReducer({}, action);
    expect(newState).toEqual({
      loading: payload
    });
  });
  it('should dispatch an action to update notification', () => {
    const payload = {
      message: '',
      user: {}
    };
    const action = {
      type: 'NOTIFICATION_SUCCESS',
      payload
    };
    const newState = UserReducer({}, action);
    expect(newState).toEqual({
      message: action.payload.message,
      user: action.payload.user,
      success: true,
    });
  });
  it('should dispatch an action when notification update fails', () => {
    const payload = {};
    const action = {
      type: 'NOTIFICATION_FAILING',
      payload
    };
    const newState = UserReducer({}, action);
    expect(newState).toEqual({});
  });
  it('should click change settings link', () => {
    document.body.innerHTML = `
    <div id="profile">
      <section id="settings">
        <form>
          <input type="time" required="true" className="control-form" name="reminderTime" id="settngs" placeholder="Enter Date" autoComplete="settings" disabled="true" />
          <button type="submit" className="submit-button button-white" >Edit</button>
        </form>
      </section>
    </div>
    `;
    myComponent.find('a#change_settings').simulate('click', {
      preventDefault: () => {
      }
     });
  });
  it('should save settings', () => {
    document.body.innerHTML =
    '<div>' +
    '  <span id="loading" />' +
    '  <span class="form_error_text" /><small></small><span>' +
    '  <button id="button" />' +
    '</div>';
    const input = myComponent.find('input#settngs');
    input.simulate('change', { target: { value: '10:00' } });
    myComponent.find('button.button-white').simulate('click', {
      preventDefault: () => {
      }
     });
  });
});
