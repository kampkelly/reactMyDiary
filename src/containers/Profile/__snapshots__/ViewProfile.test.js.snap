// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ViewProfile/> should render without throwing an error 1`] = `
<Connect(ViewProfile)
  changeSettings={[MockFunction]}
  checkNotice={[MockFunction]}
  history={
    Object {
      "push": [MockFunction],
    }
  }
  saveSettings={[MockFunction]}
  showProfile={[MockFunction]}
/>
`;
